modules/javafx.graphics/src/test/jslc/com/sun/scenario/effect/compiler/lexer/LexerBase.java

```diff
@@ -26,59 +26,66 @@
 package com.sun.scenario.effect.compiler.lexer;
 
 import com.sun.scenario.effect.compiler.JSLLexer;
-import org.antlr.runtime.ANTLRStringStream;
-import org.antlr.runtime.Token;
+import com.sun.scenario.effect.compiler.JSLParser;
+import com.sun.scenario.effect.compiler.ThrowingErrorListener;
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.Token;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
 
 import static org.junit.Assert.*;
 
 public abstract class LexerBase {
 
-    protected void assertRangeOfCharactersRecognized(char low, char high)
-        throws Exception {
-
-        for (char ch = low; ch <= high; ++ch) {
-            assertRecognized(ch);
-        }
+    protected void assertRecognized(char ch) throws Exception {
+        assertRecognized(String.valueOf(ch));
     }
 
-    protected void recognizeRange(char low, char high) throws Exception {
-        for (char ch = low; ch <= high; ++ch) {
-            recognize(ch);
+    protected void assertRecognized(String text) throws Exception {
+        Token token = recognize(text);
+        assertEquals(text, token.getText());
+
+        if (expectedTokenType() != Integer.MIN_VALUE) {
+            assertEquals(expectedTokenType(), token.getType());
         }
     }
 
-    protected void assertRecognized(char ch) throws Exception {
-        assertRecognized(String.valueOf(ch));
+    protected void assertNotRecognized(char ch) throws Exception {
+        assertNotRecognized(String.valueOf(ch));
     }
 
-    protected Token recognize(char ch) throws Exception {
-        return recognize(String.valueOf(ch));
+    protected void assertNotRecognized(String text) throws Exception {
+        assertNotRecognized(text, text);
     }
 
-    protected void assertRecognized(String text) throws Exception {
+    protected void assertNotRecognized(String text, String shouldLex) throws Exception {
         Token token = recognize(text);
-        assertEquals(text, token.getText());
+        assertEquals(shouldLex, token.getText());
 
         if (expectedTokenType() != Integer.MIN_VALUE) {
-            assertEquals(expectedTokenType(), token.getType());
+            assertFalse(expectedTokenType() == token.getType());
```

Comment:
Actually, the `assertFalse` call is opposite to `assertEquals` -- it's basically doing an "assert not equals" check, which seems odd to me, but perhaps there is a reason?

Reply:
We want to assert that it does *not* recognize the String as a certain type of token.

For example we want to test that "double" is not recognized as a `JSLLexer.TYPE`.

JUnit 4.11+ has an `assertNotEquals` method but we are not currently using a high enough JUnit version to use it.

Reply:
And, in fact, you alerted us reviewers to this change:

> Therefore instead of expecting assertRecognized to fail we added a assertNotRecognized method which checks what type of token should be lexed/parsed.

So yes, this all makes sense now.